### PR TITLE
[Model] Add multi-image input support for LLaVA-Next offline inference

### PR DESCRIPTION
**Pull request by zifeitong**
@DarkLight1337 It's good to be reviewed now. I am following #7122 and not adding a new axis to input shape as outlined in https://github.com/vllm-project/vllm/issues/4194#issuecomment-2251895609. Let me know what do you think.

- [x] Guard models that don't support multi-image input (#7126)
- [x] Support images inputs of variable sizes in the same batch
  - [x]  Support images with different number of patches (e.g. {2×2,1×{2,3,4},{2,3,4}×1})
  - [x]  Support images with different number of features

Reference: https://github.com/vllm-project/vllm/issues/4194#issuecomment-2251895609, #7122, #7392

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: tests/multimodal/test_image.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import pytest
+from transformers import AutoConfig, AutoTokenizer
+
+from vllm.multimodal.image import repeat_and_pad_image_tokens
+
+
+@pytest.mark.parametrize("model", ["llava-hf/llava-v1.6-mistral-7b-hf"])
+def test_repeat_and_pad_image_tokens(model):
```
Comment (DarkLight1337) [inline tests/multimodal/test_image.py:8]: Thanks for adding this sanity check!

File: vllm/multimodal/image.py [hunk 2]
```diff
@@ -71,28 +67,45 @@ def repeat_and_pad_image_tokens(
...
+                                    maxsplit=len(repeat_count))
+        new_prompt = ""
+        for i in range(len(repeat_count)):
+            replacement_str = "".join(
+                repeat_and_pad_token(
+                    image_token_str,
+                    repeat_count=repeat_count[i],
+                    pad_token_left=pad_token_str_left,
+                    pad_token_right=pad_token_str_right,
+                ))
```
Comment (DarkLight1337) [inline vllm/multimodal/image.py:87]:
```suggestion
        for repeat_count_item in repeat_count:
            replacement_str = "".join(
                repeat_and_pad_token(
                    image_token_str,
                    repeat_count=repeat_count_item,
                    pad_token_left=pad_token_str_left,
                    pad_token_right=pad_token_str_right,
                ))
```

**Review by DarkLight1337 — COMMENTED** [review]
Some initial comments.

**Comment by DarkLight1337** [comment]
> I am following https://github.com/vllm-project/vllm/pull/7122 and not adding a new axis to input shape as outlined in https://github.com/vllm-project/vllm/issues/4194#issuecomment-2251895609. Let me know what do you think.

We can keep this for now since it's not breaking anything. Can always revisit this later.

File: tests/models/test_llava_next.py [hunk 3]
```diff
@@ -23,7 +26,7 @@
     f"{_PREFACE} USER: <image>\nWhat is the season? ASSISTANT:",
 })
 
-models = ["llava-hf/llava-v1.6-vicuna-7b-hf"]
+models = ["llava-hf/llava-v1.6-mistral-7b-hf"]
```
[the remark this answers is not in the document]
Comment (zifeitong) [reply tests/models/test_llava_next.py:24]: Changed the prompts. Switched to mistral mainly for 32k context window.

**Comment by DarkLight1337** [comment]
VLM tests are failing

**Comment by zifeitong** [comment]
> VLM tests are failing

The issue is phi3v uses more than 1 placeholder stings (<image_1>, <image_2>, ...), and breaks the placeholder padding function. It should work after https://github.com/vllm-project/vllm/pull/7783 is merged.

**Comment by DarkLight1337** [comment]
Do you prefer to wait for #7783 or just copy the original CLIP padding code into phi3v?

**Comment by zifeitong** [comment]
> Do you prefer to wait for #7783 or just copy the original CLIP padding code into phi3v?

I looks like #7783 will be merged soon. I'll rebase my PR once it's merged and rerun the tests.

**Review by DarkLight1337 — APPROVED** [review]
Thanks again for implementing this!

**Pull request merged**